draft-specifications/CapTP Specification.md

```diff
@@ -0,0 +1,1002 @@
+the promise. If multiple values are pipelined on, without the pick operation,
+the resulting promise MUST be broken with an error and not fulfilled.
+
+### Promise pipelining by example
+
+In this example we send an initial message to the object exported at position
+`5` (a factory builder) and tell the remote session to export the promise at
+answer position `3` (a car factory), then we send a message to the promise at
+that answer position to make a car and export it at answer position `4`, finally
+we send a message to the promise of a car at answer position `4` to drive it.
+
+The messages delivered are as follows:
+
+```
+<op:deliver <desc:export 5> ['make-car-factory] 3 <desc:import-object 15>>
+<op:deliver <desc:answer 3> ['make-car] 4 <desc:import-object 16>>
+<op:deliver <desc:answer 4> ['drive] 5 <desc:import-object 17>>
+```
+
+In this way, it is not necessary to wait to dispatch messages until the promise
+has been resolved. Messages can just be sent in advance to a promise of the
+eventual object as they normally would to its eventual resolution. This also
+improves efficiency by reducing the number of round trips needed to perform the
+same task. Take the above example of a car factory which produces a car, the car
+then can be driven which will produce a noise. Without promise pipelining this
+would look something like this:
+
+1.  Send message to create car factory
+2.  Get back a reference to the car factory once it is made
+3.  Send message to car factory to make a car
+4.  Get back a reference to the car
+5.  Send message to car to drive
+6.  Get back the driving noise.
+
+This is compared to when you use promise pipelining:
+
+1.  Send message to create car factory
+2.  Send message to car factory promise to make a car
+3.  Send message to car promise to drive
+4.  Get back driving noise
+
+The other CapTP session, which is making the car factory, car, and finally car
+noise, is able deliver those messages immediately to the local objects on their
+side as they become fulfilled and send the car noise back at the end.
+
+
+# [Cryptography](#cryptography)
+
+Each CapTP session has a per session key pair which is used for signing certain
+structures for example in the [third party handoffs
+section](#third-party-handoffs). These key pair values are generated EdDSA with
+a SHA512 hash.
+
+**NOTE:** These representations are considered temporary and we are anticipating
+replacing them, probably with record-based representations.
+
+## [Public key](#public-key)
+
+Public keys are formatted based on gcrypt's s-expression format, using EdDSA
+public keys and the SHA512 hash algorithm. The EdDSA public keys are based on
+the Ed25519 elliptic curve. The public key is formatted as follows:
+
+```text
+(public-key (ecc (curve Ed25519) (flags eddsa) (q ...) (s ...)))
+```
+
+Everything except the `...` must be verbatim. The `...` is replaced by Binary
+Data typed values representing the public key.
```

Comment:
This representation is based off the s-expression format that gcrypt uses, so this exact expression (except with `q` and `s` values filled with the public key data). The expression is actually a Syrup sequence composed of symbols. Spritely is not attached to this format and we probably would be looking to change it within the future.

I think this might be a good candidate for it's own `desc` struct, but I think this could be best discussed within the group to see what works best for everyone here.

Reply:
I've improved the text describing the two cryptography values, hopefully it's a bit clearer now.